examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,2010 @@
+                        noisy_teacher_input.detach(),
+                        teacher_timesteps,
+                        encoder_hidden_states=prompt_embeds,
+                    ).sample
+
+                    teacher_uncond_noise_pred = teacher_unet(
+                        noisy_teacher_input.detach(),
+                        teacher_timesteps,
+                        encoder_hidden_states=uncond_prompt_embeds,
+                    ).sample
+
+                    # Get the teacher's CFG estimate of x_0.
+                    teacher_cfg_noise_pred = w * teacher_cond_noise_pred + (1 - w) * teacher_uncond_noise_pred
+                    teacher_x_0 = denoiser(teacher_cfg_noise_pred, teacher_timesteps, noisy_teacher_input)
+
+                ############################
+                # 9. Discriminator Loss
+                ############################
+                discriminator_optimizer.zero_grad(set_to_none=True)
+
+                # 1. Decode real and fake (generated) latents back to pixel space.
+                # NOTE: the paper doesn't mention this explicitly AFAIK but I think this makes sense since the
+                # pretrained feature network for the discriminator operates in pixel space rather than latent space.
+                unscaled_student_x_0 = (1 / vae.config.scaling_factor) * student_x_0
+                # Perform batched decode with batch size of at most args.vae_encode_batch_size
+                student_gen_image = []
+                for i in range(0, unscaled_student_x_0.shape[0], args.vae_encode_batch_size):
+                    student_gen_image.append(
+                        vae.decode(
+                            unscaled_student_x_0[i : i + args.vae_encode_batch_size].to(dtype=weight_dtype)
+                        ).sample
+                    )
+                student_gen_image = torch.cat(student_gen_image, dim=0)
+
+                # 2. Get discriminator real/fake outputs on the real and fake (generated) images respectively.
+                disc_output_real = discriminator(pixel_values.float(), text_embedding, image_embedding)
+                disc_output_fake = discriminator(student_gen_image.detach().float(), text_embedding, image_embedding)
+
+                # 3. Calculate the discriminator real adversarial loss terms.
+                d_logits_real = disc_output_real.logits
+                # Use hinge loss (see section 3.2, Equation 3 of paper)
+                d_adv_loss_real = torch.mean(F.relu(torch.ones_like(d_logits_real) - d_logits_real))
+
+                # 4. Calculate the discriminator R1 gradient penalty term with respect to the gradients from the real
+                # data.
+                d_r1_regularizer = 0
+                for k, head in discriminator.heads.items():
+                    head_grad_params = torch.autograd.grad(
+                        outputs=d_adv_loss_real, inputs=head.parameters(), create_graph=True
```

Comment:
according to the paper, the r1 penalty seems to be computed w.r.t head's input instead of head's parameters?

Reply:
Thanks for the catch, I think you're right. In fact there seem to be several problems in the current implementation:

- As you pointed out, the gradient penalty should be calculated with respect to the discriminator head inputs. That's not available in the current code, but if the input to discriminator head `k` is available as `features[k]`, I think the fix would be to set the `inputs` argument to [`torch.autograd.grad`](https://pytorch.org/docs/stable/generated/torch.autograd.grad.html#torch.autograd.grad) to `features[k]` (if I understand autograd correctly).
- It looks like I misunderstood the definition of the R1 gradient penalty. The [ADD paper](https://arxiv.org/pdf/2311.17042.pdf) cites [this paper](https://proceedings.mlr.press/v80/mescheder18a/mescheder18a.pdf) when discussing the R1 gradient penalty, and the latter paper defines the R1 gradient penalty as

```math
R_1(\phi) = \frac{1}{2}\mathbb{E}_{x_0 \sim p_{data}(x)}[||\nabla D_{\phi, k}(F_k(x_0))||^2]
```

So it seems like we should be using the L2 norm rather than the L1 norm when calculating the gradient penalty. It's also possible that the implementation is off by factor of $`\frac{1}{2}`$; I'm not sure if they absorb the factor of $`\frac{1}{2}`$ into $`\gamma`$ (and their reported value of $`\gamma = 10^{-5}`$).

@patil-suraj @sayakpaul does this sound correct to you guys?

Reply:
yeah, regarding formula it is l2 norm, something like this
```
d_r1_regularizer = sum((torch.linalg.vector_norm(grad.view(grad.size(0), -1), dim=1) ** 2).mean() for grad in feature_grads)
```

Reply:
I have added a tentative fix for the discriminator R1 gradient penalty for the SD ADD script in commit ab46142. In particular this part

https://github.com/huggingface/diffusers/blob/ab46142de967b21e031e2f7cdac82898e8a7e5f9/examples/add/train_add_distill_sd_wds.py#L1860-L1862

feels weird to me but seems necessary because the `feature`s in `features_real` don't usually have gradients because the `feature_network` is frozen. @erliding @patil-suraj @sayakpaul would be great if you could looks this over

Reply:
yep, you need to explicitly call `feature.requires_grad_()` before passing them to heads

Reply:
@erliding's inputs seem correct to me. 

One silly question: how important is gradient penalty for ADD?

Reply:
@sayakpaul the authors report that it is pretty useful:

<img width="300" alt="add_paper_disc_grad_penalty" src="https://github.com/huggingface/diffusers/assets/58458699/1882c356-a8aa-4aae-a782-6a7700c2f539">

(from Section 3.2 of the [ADD paper](https://arxiv.org/pdf/2311.17042.pdf))

Reply:
@sayakpaul as far as my understanding, the gradient penalty could help enforce Lipschitz continuity (thus gradient of output w.r.t input) on discriminator which is a requirement of Wasserstein Gan

Reply:
Right, but we aren't using the earth mover distance here in the loss no?

Reply:
yeah, it's hinge loss here :) gradient penalty should bring similar benefits though

Reply:
@dg845 It does look good to me. We need to enable grads for feature inputs to be able to compare the grad penalty, if the input does not have grad enabled I think `create_graph=True` will   complain. 

for reference, DDGAN training has this
- https://github.com/NVlabs/denoising-diffusion-gan/blob/main/train_ddgan.py#L338
- https://github.com/NVlabs/denoising-diffusion-gan/blob/main/train_ddgan.py#L351